README.rst
Yet Another YubiKey OTP Validation Server
=========================================

Several other implementations are available. Some of them are not secure enough:

 * `YubiServe <https://code.google.com/p/yubico-yubiserve>`_ (Python). `SQL
   injections
   <https://code.google.com/p/yubico-yubiserve/issues/detail?id=38>`_,
 * `yubiserver <http://www.include.gr/debian/yubiserver/>`_ (C). SQL injections
   (CVE-2015-0842), buffer overflows (CVE-2015-0843).

Official implementation is written in PHP (sigh...), and I don't know Go enough
to audit digintLab's implementation:

 * `yubikey-val <https://developers.yubico.com/yubikey-val/>`_ (PHP) by Yubico,
 * `yubikey-server <https://github.com/digintLab/yubikey-server>`_ (Go).

This is a complete rewrite of `YubiServe
<https://code.google.com/p/yubico-yubiserve>`_ because the original project
seems not to be designed with security in mind. Copy-and-paste programming made
code reviews nearly impossible, there is no protection against SQL injection,
etc.

This fork was given a new name to make it easy for people to differentiate from
the original project.


Usage
=====

Create a new database::

    $ ./tools/dbcreate.py ./yubikeys.sqlite3

Plug and flash the YubiKeys (keys are also written to the database)::

    $ ./tools/flash.py gbush ./yubikeys.sqlite3
    $ ./tools/flash.py bobama ./yubikeys.sqlite3

Add a new API key (here, the API key name is ``developers``)::

    $ ./tools/dbconf.py -aa developers ./yubikeys.sqlite3

Run the server::

    $ ./src/yubiserve.py --db ./yubikeys.sqlite3

That's it. The servers wanting to make use of two factor authentication need to
be configured. The following paragraph shows an example for OpenSSH.


OpenSSH configuration example
=============================

Here's a summary of `Yubico's documentation
<https://developers.yubico.com/yubico-pam/Yubikey_and_SSH_via_PAM.html>`_.

Get information about users and API on the machine hosting
``yubikeys.sqlite3``::

    $ ./tools/dbconf.py -yl ./yubikeys.sqlite3
    2 keys into database:
    [Nickname]              >> [PublicID]            >> [Active]
    gbush                   >> ibhdhehrhkhuifhv      >> 1
    bobama                  >> ibibhdhvhdhbhthb      >> 1
    
    $ ./tools/dbconf.py -al ./yubikeys.sqlite3
    1 keys into database:
    [Id]    >> [Keyname]            >> [Secret]
    1       >> developers           >> ckFsWU5scVNXRjVZc3lJUmpIVzU=

On the OpenSSH machine, add users to ``/etc/yubimap``::

    $ cat /etc/yubimap
    barack:ibibhdhvhdhbhthb
    george:ibhdhehrhkhuifhv

Configure PAM to use YubiKey authentication (take care of API ``id`` and API
``key`` values)::

    $ head /etc/pam.d/sshd | grep include
    #@include common-auth
    @include yubi-auth
    
    $ cat /etc/pam.d/yubi-auth
    auth       required     pam_yubico.so authfile=/etc/yubimap id=1 key=ckFsWU5scVNXRjVZc3lJUmpIVzU= url=http://yubikeyval.local:8000/wsapi/2.0/verify?id=%d&otp=%s mode=client token_id_length=16

Configure OpenSSH::

    $ tail -4 /etc/ssh/sshd_config
    ChallengeResponseAuthentication  no
    Match User george,barack
        PasswordAuthentication       yes
        AuthenticationMethods        publickey,password


TODO
====

OATH/HOTP is not supported at present.


Original author
===============

 * Alessio Periloso <mail *at* periloso.it>


debian/changelog
yubikeyedup (0.1-1) unstable; urgency=low

  * source package automatically created by stdeb 0.8.2

 -- jaroug <jaroug+yubikeyedup@gmail.com>  Thu, 06 Apr 2017 16:29:15 +0000


src/yubiserve.py
#!/usr/bin/env python

import BaseHTTPServer
import SocketServer
import base64
import hashlib
import hmac
import optparse
import os
import re
import signal
import socket
import sys
import threading
import time
import urllib
import urlparse

import yubistatus
import validate
import html
from sql import *


class YubiServeHandler:
    def __init__(self, sql_connection, params, vclass):
        self.sql = SQL(sql_connection)
        self.params = params
        self.vclass = vclass

    def sign_message(self, answer, api_key):
        data = [ '%s=%s' % (k, v) for (k, v) in answer.iteritems() ]
        data.sort()
        data = '&'.join(data)

        otp_hmac = hmac.new(api_key, str(data), hashlib.sha1)
        otp_hmac = base64.b64encode(otp_hmac.digest())

        return otp_hmac

    def build_answer(self, status, answer, api_key=''):
        answer['status'] = status
        answer['h'] = self.sign_message(answer, api_key)

        data = '\r\n'.join([ '%s=%s' % (k, v) for (k, v) in answer.iteritems() ])
        data += '\r\n'

        return data

    def do_validate(self):
        answer = { 't': time.strftime("%Y-%m-%dT%H:%M:%S"), 'otp': '' }

        # API id and OTP are required
        if not self.params.has_key('id') or not self.params.has_key('otp'):
            return self.build_answer(yubistatus.MISSING_PARAMETER, answer)

        # ensure API id is valid
        if not self.sql.select('get_api_secret', [self.params['id']]):
            return self.build_answer(yubistatus.NO_SUCH_CLIENT, answer)

        api_key = base64.b64decode(self.sql.result[0])

        # do token validation
        vclass = self.vclass(self.sql)
        status = vclass.set_params(self.params, answer)
        if status == yubistatus.OK:
            status = vclass.validate()
        return self.build_answer(status, answer, api_key)


class YubiHTTPServer(BaseHTTPServer.BaseHTTPRequestHandler):
    __base = BaseHTTPServer.BaseHTTPRequestHandler
    __base_handle = __base.handle

    server_version = 'YubiKeyedUp/1.0'
    vclasses = {
        '/wsapi/2.0/verify': validate.Yubico,
        '/wsapi/2.0/oathverify': validate.OATH,
    }

    PARAM_REGEXP = {
        'id': '[0-9]{1,3}',
        'otp': '[cbdefghijklnrtuv]{0,16}[cbdefghijklnrtuv]{32}',
        'publicid': '[cbdefghijklnrtuv]{0,16}',
        'nonce': '[a-zA-Z0-9]{16,40}',
    }

    def __init__(self, request, client_address, server):
        global sqlite_db
        self.sql_connection = connect_to_db(sqlite_db)
        return BaseHTTPServer.BaseHTTPRequestHandler.__init__(self, request, client_address, server)

    def setup(self):
        self.connection = self.request
        self.rfile = socket._fileobject(self.request, 'rb', self.rbufsize)
        self.wfile = socket._fileobject(self.request, 'wb', self.wbufsize)

    def getToDict(self, qs):
        dict = {}
        for singleValue in qs.split('&'):
            if not '=' in singleValue:
                continue
            key, value = singleValue.split('=', 1)
            value = urllib.unquote_plus(value)
            if self.PARAM_REGEXP.has_key(key) and re.match(self.PARAM_REGEXP[key], value):
                dict[key] = value
        return dict

    def do_GET(self):
        url = urlparse.urlparse(self.path, 'http')

        if self.vclasses.has_key(url.path):
            params = self.getToDict(url.query)
            vclass = self.vclasses[url.path]
            handler = YubiServeHandler(self.sql_connection, params, vclass)
            data = handler.do_validate()
            content_type = 'text/plain'
        else:
            data = index = html.index
            content_type = 'text/html'

        self.send_response(200)
        self.send_header('Content-type', content_type)
        self.end_headers()
        self.wfile.write(data)


class ThreadingHTTPServer(SocketServer.ThreadingMixIn, BaseHTTPServer.HTTPServer):
    pass


def stop_signal_handler(signum, frame):
    yubiserveHTTP.shutdown()
    sys.exit(0)


def main():
    parser = optparse.OptionParser('Usage: %prog [options]')
    parser.add_option('-d', '--db', default='./yubikeys.sqlite3', dest='db')
    parser.add_option('-a', '--address', default='0.0.0.0', dest='host')
    parser.add_option('-p', '--port', default='8000', dest='port')
    (options, args) = parser.parse_args()
    global sqlite_db
    sqlite_db = options.db

    yubiserveHTTP = ThreadingHTTPServer((options.host, int(options.port)), YubiHTTPServer)

    signal.signal(signal.SIGINT, stop_signal_handler)

    http_thread = threading.Thread(target=yubiserveHTTP.serve_forever)
    http_thread.setDaemon(True)
    http_thread.start()

    signal.pause()


if __name__ == '__main__':
    main()
